### PR TITLE
Support upsert in Postgres temp tables

### PR DESCRIPTION
**Pull request by rschellhorn**
### Detail

This Pull Request changes the Postgres database adapter. It allows me to create models that are backed by a temporary table and use them just like regular tables. Without this change e.g. `insert_all` and `upsert_all` are not possible.

### Test

```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  gem "activerecord", path: './activerecord'
  gem "pg"
end

require "active_record"
require "minitest/autorun"
require "logger"

ActiveRecord::Base.establish_connection(adapter: "postgresql", url: "postgres://localhost/rails-issue") rescue puts 'Please create a DB "rails-issue" yourself first'
ActiveRecord::Base.logger = Logger.new(STDOUT)

class Post < ActiveRecord::Base
  def self.with_temp_table(as:)
    transaction do
      connection.create_table(table_name, temporary: true, force: true, as: as)
      yield
    end
  end
end

class BugTest < Minitest::Test
  def test_upsert_in_temp_table

    # Create a model backed by a temp table. Prefill the table with some data from a query.
    Post.with_temp_table(as: "select 1 as id, 'foo' as some_column") do
      # Upserts require a unique constraint
      Post.connection.add_index(Post.table_name, :id, unique: true)

      assert_equal Post.new(id: 1, some_column: 'foo'), Post.sole

      # The following line fails on master with:
      #
      #   ActiveRecord::StatementInvalid: PG::SyntaxError: ERROR:  syntax error at or near ")"
      #   LINE 1: ...id","some_column") VALUES (1, 'bar') ON CONFLICT () DO UPDAT...
      #
      # The temp table, including all indices, constraints, etc are not 'seen'
      # by Rails. The PostgreSQL adapter only considers stuff in the schemas
      # returned by current_schemas(false). The temporary namespace
      # 'pg_temp_xyz' is excluded.
      #
      # See https://www.postgresql.org/docs/15/functions-info.html
      #
      #   current_schemas ( include_implicit boolean ) → name[]
      #
      #   Returns an array of the names of all schemas presently in the
      #   effective search path, in their priority order. (Items in the current
      #   search_path setting that do not correspond to existing, searchable
      #   schemas are omitted.) If the Boolean argument is true, then
      #   implicitly-searched system schemas such as pg_catalog are included in
      #   the result.
      #
      # If you include implicit tables, via current_schemas(true), the
      # temporary table is found, but that is also true for the other system
      # schemas that we don't want.
      #
      # Luckily Postgres provides a means to find the temp schema name via:
      #
      #   pg_my_temp_schema () → oid
      #
      #   Returns the OID of the current session's temporary schema, or zero if
      #   it has none (because it has not created any temporary tables).
      #
      # If we add that to the search path, then temp tables work just like
      # regular tables.
      Post.upsert_all([{ id: 1, some_column: 'bar' }], unique_by: :id)

      assert_equal Post.new(id: 1, some_column: 'bar'), Post.sole
    end
  end
end
```

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [ ] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

### CODE REVIEW

**Comment by yahonda** [comment]
Thanks for the update. 

* Query updated by this pull request

```sql
SELECT c.relname FROM pg_class c LEFT JOIN pg_namespace n ON n.oid = c.relnamespace WHERE n.nspname = ANY (array_prepend((SELECT pg_my_temp_schema()::regnamespace::name),current_schemas(false))) AND c.relkind IN ('r','v','m','p','f')
```

* Current query

```sql
SELECT c.relname FROM pg_class c LEFT JOIN pg_namespace n ON n.oid = c.relnamespace WHERE n.nspname = ANY (current_schemas(false)) AND c.relkind IN ('r','v','m','p','f')
```

I recall that as of right now, the main Rails branch must support PostgreSQL 9.3.

https://github.com/rails/rails/blob/70e74216d7556dc8b68841814fb8accafb4e6cbf/activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb#L601-L605

However, since PostgreSQL 9.5, the `regnamespace' type has been introduced. 
Therefore, it will require one of these two options:

- Replace `regnamespace` with something supported by PostgreSQL 9.3.
- Bump the PostgreSQL version requirements to 9.5 or higher.

Actually, I opened this pull request to bump the minimum version of PostgreSQL https://github.com/rails/rails/pull/41526, which was closed. I think it is worth reopening it to get some more feedback if the option 1 is not suitable.

**Comment by rschellhorn** [comment]
Thanks for the feedback. I did not consider a minimum Postgresql version.

> Replace regnamespace with something supported by PostgreSQL 9.3.

I'll look into that. There probably is something available.

**Comment by matthewd** [comment]
PG version/API considerations aside, it seems like this would be Very Bad News for the schema cache. I don't believe we even have it account for the fact that newly-created tables can disappear during a rollback, let alone the more creative ways that the dynamic behaviour of temp tables can complicate things... they'll be persistent as long as the connection is active (despite it passing through the pool, and thus moving between requests); they'll be present on one connection but not another (the schema cache is shared within a pool, not unique per connection); you could end up with two codepaths creating differently-shaped but same-named temp tables.

Basically, while I don't think this is currently deliberately broken, and it also sounds like it mostly sorta-kinda works... I do think the underlying behaviour (accessing a temp table from an AR model) seems likely to be a generally bad idea, and thus not something we should encourage, given all the other limitations.

**Comment by rschellhorn** [comment]
Thank you for the feedback.

> you could end up with two codepaths creating differently-shaped but same-named temp tables.

As in, this is a result of the global schema cache in Rails you mean? In Postgresql this is also an issue for applications that use a connection pool of course. It solves it by offering a means to scope the temp table to a transaction instead of the connection. You can pass in the option `ON COMMIT DROP`. Rails can do the same:

```ruby
TempModel.with_temp_table(as: "select 1 as id, 'foo' as some_column", option: 'ON COMMIT DROP') { ... }
```

I see how this is really complicated to work around with a global schema cache. My PR certainly won't cut it.

> I do think the underlying behaviour (accessing a temp table from an AR model) seems likely to be a generally bad idea, and thus not something we should encourage, given all the other limitations.

In most cases I would agree. There are however use cases that could benefit. Mine is one imho:
- we are annotating many records
- there are too many rows to load in memory, so I really like to keep the data in the DB. They need to be annotated in one go unfortunately; they are dependent.
- I don't want to put a long running transaction on the source table
- So instead, I first copy the data in a temp, process it (many subsequent upserts) and finally write the result back to the source table.

This indeed is perhaps a niche use case Active Record doesn't have to support. Falling back to raw sql is a good enough alternative for me.

So, thanks you again for reviewing it and put my change in perspective. I am fine with closing this PR.
